### PR TITLE
Add `DataStream#getIndicesPastRetention`

### PR DESCRIPTION
**Pull request by andreidan**
This adds the `getIndicesPastRetention` method on the `DataStream` 
itself to simplify the logic in `DataLifecycleService`

Relates to #93596

### CODE REVIEW

**Comment by andreidan** [comment]
Restarting CI due to https://github.com/elastic/elasticsearch/issues/94096

**Comment by andreidan** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 1]
```diff
@@ -284,7 +286,7 @@ public boolean isHidden() {
 
     /**
      * Determines whether this data stream is replicated from elsewhere,
-     * for example a remote cluster.
+     * for example a remote cluster
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:289]: Spotless is cheeky :)

File: modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java [hunk 2]
```diff
@@ -188,47 +189,30 @@ private void maybeExecuteRollover(ClusterState state, DataStream dataStream) {
...
-            for (int i = 0; i < backingIndices.size() - 1; i++) {
-                IndexMetadata backingIndex = state.metadata().index(backingIndices.get(i));
+            Metadata metadata = state.metadata();
+            List<Index> backingIndicesOlderThanRetention = dataStream.getIndicesPastRetention(metadata::index, nowSupplier);
+
+            for (Index index : backingIndicesOlderThanRetention) {
+                IndexMetadata backingIndex = metadata.index(index);
                 if (backingIndex == null || isManagedByDLM(dataStream, backingIndex) == false) {
                     continue;
                 }
```
Comment (dakrone) [inline modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java:199]: Should the `isManagedByDLM` call be part of the `getIndicesPastRetention` also?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:289]: Spotless hates correct grammar??

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 3]
```diff
@@ -572,6 +574,56 @@ public DataStream snapshot(Collection<String> indicesInSnapshot) {
         );
     }
 
+    public List<Index> getIndicesPastRetention(Function<String, IndexMetadata> indexMetadataSupplier, LongSupplier nowSupplier) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:577]: Can you add javadocs for this method?

**Review by dakrone — APPROVED** [review]
This looks good to me, I added a question about moving the management check into the data stream also, as well as a request for javadocs

**Pull request merged**